pkg/agent/reconcile_force_io_error_pods.go

```diff
@@ -0,0 +1,113 @@
+package agent
+
+import (
+	"context"
+	"fmt"
+
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/events"
+	"k8s.io/klog/v2"
+
+	"github.com/piraeusdatastore/piraeus-ha-controller/pkg/eviction"
+	"github.com/piraeusdatastore/piraeus-ha-controller/pkg/metadata"
+)
+
+type forceIoErrorReconciler struct {
+	opt    *Options
+	client kubernetes.Interface
+}
+
+var _ Reconciler = &forceIoErrorReconciler{}
+
+// NewForceIoErrorReconciler creates a reconciler that evicts pods if a volume is reporting IO errors.
+//
+// If DRBD is in "force IO failures" mode, all opener processes will see IO errors. This lasts until all openers
+// closed the DRBD device, at which point DRBD will start behaving normally again. In order for all openers to be
+// closed we need to force all local Pods to be stopped. This is what this reconciler does:
+// * Adding a taint on the node, causing new Pods to avoid the node.
+// * Evicting all pods using that volume from the failed node, creating new pods to replace them.
+func NewForceIoErrorReconciler(opt *Options, client kubernetes.Interface) Reconciler {
+	return &forceIoErrorReconciler{
+		opt:    opt,
+		client: client,
+	}
+}
+
+func (f *forceIoErrorReconciler) RunForResource(ctx context.Context, req *ReconcileRequest, recorder events.EventRecorder) error {
+	if !req.Resource.ForceIoFailures {
+		klog.V(4).Infof("resource '%s' is not forcing io-errors", req.Resource.Name)
+		return nil
+	}
+
+	// NB: we use runtime.Object here, so we don't accidentally create a "nil interface value", which would
+	// confuse the event-recorder into panicking.
+	var matchingVA runtime.Object
+	for _, va := range req.Attachments {
+		if va.Spec.NodeName == f.opt.NodeName {
+			matchingVA = va
+			break
+		}
+	}
+
+	var localPods []*corev1.Pod
+	for _, pod := range req.Pods {
+		if pod.Spec.NodeName != f.opt.NodeName {
+			klog.V(4).Infof("Pod '%s/%s' not running on this node", pod.Namespace, pod.Name)
+			continue
+		}
+
+		if len(pod.ObjectMeta.OwnerReferences) == 0 {
+			klog.V(4).Infof("Pod '%s/%s' does not have an owner", pod.Namespace, pod.Name)
+			continue
+		}
+
+		if _, exists := pod.ObjectMeta.Annotations[metadata.AnnotationIgnoreFailOver]; exists {
+			klog.V(4).Infof("Pod '%s/%s' is exempt from automatic fail over", pod.Namespace, pod.Name)
+			continue
+		}
+
+		localPods = append(localPods, pod)
+	}
+
+	if len(localPods) == 0 {
+		klog.V(2).Infof("No pods to evict")
+	}
```

Comment:
but you still want to continue? no early return or anything, just a info?

Reply:
Good catch, should early return.